crates/ruma-state-res/src/event_auth.rs

```diff
@@ -497,10 +497,10 @@ fn valid_membership_change(
         (target_membership == MembershipState::Join).then_some(&power_levels.users_default)
     });
 
-    let mut join_rules = JoinRule::Invite;
-    if let Some(jr) = &join_rules_event {
-        join_rules = from_json_str::<RoomJoinRulesEventContent>(jr.content().get())?.join_rule;
-    }
+    // Instead of failing if the rules can't be parsed, assume invite-only
+    let join_rules = join_rules_event
+        .and_then(|x| from_json_str::<RoomJoinRulesEventContent>(x.content().get()).ok())
+        .map_or(JoinRule::Invite, |x| x.join_rule);
```

Comment:
On the related issue, I quote this section of the spec:

> If the room is `restricted` but no valid conditions are presented then the room is effectively invite only.

&mdash; https://spec.matrix.org/v1.10/client-server-api/#restricted-rooms

Also on the related issue, I explain that:

> The user facing impact of [failing to parse the join rules] is that it's impossible to join these rooms even if the user gets invited to such a room.

In order to be robust against this, this change is required, and it seems to be what the spec says to do: the conditions are invalid (don't parse) and so it should be treated as invite only.

Reply:
Ah, yes, you're right, I need to make this behavior more localized to restricted rooms.

Reply:
That's exactly what the spec says though. If there are no valid join conditions, it should behave as `invite`. Otherwise anyone can join the room, which is clearly not what `restricted` is for.

Reply:
Ah right, I see. I believe that you are right. The `join_rule` should stay as `restricted` or `knock_restricted` to allow it to pass the auth rules..